backend/src/authentication/handlers.rs

```diff
@@ -0,0 +1,261 @@
+    cookie.set_secure(SECURE_COOKIES);
+    cookie.set_same_site(SameSite::Strict);
+}
+
+/// Login endpoint, authenticates a user and creates a new session + session cookie
+#[utoipa::path(
+    post,
+    path = "/api/user/login",
+    request_body = Credentials,
+    responses(
+        (status = 200, description = "The logged in user id and user name", body = LoginResponse),
+        (status = 401, description = "Invalid credentials", body = ErrorResponse),
+        (status = 500, description = "Internal server error", body = ErrorResponse),
+    ),
+)]
+pub async fn login(
+    State(users): State<Users>,
+    State(sessions): State<Sessions>,
+    jar: CookieJar,
+    Json(credentials): Json<Credentials>,
+) -> Result<impl IntoResponse, APIError> {
+    let Credentials { username, password } = credentials;
+
+    // Check the username + password combination
+    let user = users.authenticate(&username, &password).await?;
+
+    // Remove old sessions, we do this after a login to prevent the necessity of periodical cleanup jobs
+    sessions.delete_old_sessions().await?;
+
+    // Create a new session and cookie
+    let session = sessions.create(user.id(), SESSION_LIFE_TIME).await?;
+
+    // Add the session cookie to the response
+    let mut cookie = session.get_cookie();
+    set_default_cookie_properties(&mut cookie);
+    let updated_jar = jar.add(cookie);
+
+    Ok((updated_jar, Json(LoginResponse::from(&user))))
+}
+
+/// Logout endpoint, deletes the session cookie
+#[utoipa::path(
+    post,
+    path = "/api/user/logout",
+    responses(
+        (status = 200, description = "Successful logout, or user was already logged out"),
+        (status = 500, description = "Internal server error", body = ErrorResponse),
+    ),
+)]
+pub async fn logout(
+    State(sessions): State<Sessions>,
+    jar: CookieJar,
+) -> Result<impl IntoResponse, APIError> {
+    let Some(mut cookie) = jar.get(SESSION_COOKIE_NAME).cloned() else {
+        // no cookie found, return OK
+        return Ok((jar, StatusCode::OK));
+    };
+
+    // Remove session form the database
+    let Some(session) = sessions.get_by_key(cookie.value()).await? else {
+        // no session found, return OK
+        return Ok((jar, StatusCode::OK));
+    };
+
+    sessions.delete(session.session_key()).await?;
+
+    // Set cookie parameters, these are not be retrieved by the request, and have to match in order to clear the cookie
+    set_default_cookie_properties(&mut cookie);
+    let updated_jar = jar.remove(cookie);
+
+    Ok((updated_jar, StatusCode::OK))
+}
+
+#[cfg(test)]
+mod tests {
+    use axum::{body::Body, http::Request, routing::post, Router};
+    use http_body_util::BodyExt;
+    use hyper::{header::CONTENT_TYPE, Method};
+    use sqlx::SqlitePool;
+    use tower::ServiceExt;
+
+    use super::{login, logout};
+    use crate::{
+        authentication::{Credentials, LoginResponse},
+        AppState,
+    };
+
+    fn create_app(pool: SqlitePool) -> Router {
```

Comment:
These tests look a lot like integration tests, maybe they should be moved to the `backend/tests/` folder

Reply:
I see, in other modules we don't build the `Router` and just use the handlers directly. Not sure what's better though.

Reply:
I think using the tower oneshot method is nice since you could also test any middleware, but more importantly, also tests the request extractors.

Reply:
> I think using the tower oneshot method is nice since you could also test any middleware, but more importantly, also tests the request extractors.

Agreed, I think we can use this method for more tests.